### DISCUSSION TITLE
How can I use vue3-sfc-loader with vuetify?

### DISCUSSION THREAD

**Discussion by tve** [FranckFreiburger/vue3-sfc-loader#85] [Q&A] [answered]
I'm hitting exactly the issue described in the closed issue #82, but that issue was closed by the author without a solution...
More specifically, using any vuetify component results in an error of the form `Unknown custom element: <v-text-field> - did you register the component correctly? For recursive components, make sure to provide the "name" option`.

~Update: oddly, the component that produced the error works fine in the end..~ (not true)

**Comment by FranckFreiburger** [comment]
(need an answer to close this entry)

**Comment by tve** [comment]
The problem with Vuetify is that most bundlers do auto-import of the components. The result is that the vuetify components don't get registered globally with vue and that's why they're not found when referenced by a dynamically loaded SFC.

Even if the components were globally registered, which ones would actually be available would depend on what the existing app into which an SFC component is loaded already makes use of (I'm loading components into an app as opposed to dynamically loading the full app). So what I'm inclined to do is to either figure out how to load vuetify fully and register all its components globally, or have a 'whitelist' and explicitly load those.

**Comment by adamjimenez** [comment]
Did you get anywhere with this? I'm trying to do the same thing with Vuetify 3.

**Reply by tve** [reply]
it's been a while... memory fades...
There's a fundamental problem, which is to decide which components of Vuetify get linked into your app. The dynamically loaded components will only have access to those. I ended up heavily using es-module-shims to do dynamic loading: generate a blob from the SFC, generate a URL for the blob, and ask es-module-shims to load it. The last two steps are captured in https://github.com/flexdash/flexdash/blob/main/src/utils/map-import.js
What this allows is to use the same import statement in the dyn loaded stuff as you would normally. However, it doesn't magically pull new pieces of Vuetify from a server.
The code that calls the above utility function is https://github.com/flexdash/flexdash/blob/main/src/widgets/custom-widget.vue#L106-L134 but it doesn't do the compilation step, that is handled server-side in a node app to avoid loading the sfc compiler into the browser.

**Comment by adamjimenez** [comment]
Thanks. Most of that went over my head. I figured out how to get components working, guess I can just manually include them.
It feels a bit dirty but is simple and seems to work thus far.

```vue
 <template>
    <VBtn text="Label"></VBtn>
  </template>
  
  <script>
    import Vuetify from 'https://cdn.jsdelivr.net/npm/vuetify@3.3.23/dist/vuetify-labs.min.js';
    let VBtn = Vuetify.components.VBtn
  
    export default {
      components: {
        VBtn
      },
    }
  </script>
```

**Comment by FranckFreiburger** [comment, marked as answer]
Hello tve,
you can try this minimalist Vuetify example:

```html
<!DOCTYPE html>
<html>
  <head>
    <link rel="shortcut icon" href="data:image/x-icon;," type="image/x-icon"> 
    <script src="https://cdn.jsdelivr.net/npm/vue@3.4.15/dist/vue.runtime.global.prod.js"></script>

    <script src="https://cdn.jsdelivr.net/npm/vuetify@3.5.1/dist/vuetify.min.js"></script>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/vuetify@3.5.1/dist/vuetify.min.css">

    <script src="https://cdn.jsdelivr.net/npm/vue3-sfc-loader/dist/vue3-sfc-loader.js"></script>
  </head>
<body>

<script>

  // <!--
  const config = {
    files: {
      '/main.vue': `
        <template>

          <v-btn>
            my button
          </v-btn>
          
        </template>
      `,
    }
  };
  // -->
  

  const options = {
    devMode: true,
    moduleCache: {
      vue: Vue,
    },
    async getFile(url) {
      
      // for convenience, example files a loaded from strings
      if ( config.files[url] )
        return config.files[url];

      // other files are loaded as usual, over the network
      const res = await fetch(url);
      if ( !res.ok )
        throw Object.assign(new Error(res.statusText + ' ' + url), { res });
      return {
        getContentData: asBinary => asBinary ? res.arrayBuffer() : res.text(),
      }
    },
    addStyle(styleStr) {

      const style = document.createElement('style');
      style.textContent = styleStr;
      const ref = document.head.getElementsByTagName('style')[0] || null;
      document.head.insertBefore(style, ref);
    },

    log(type, ...args) {
      console[type](...args);
    },
  }

  const vuetify = Vuetify.createVuetify({
  // ... your configuration
  })

  window['vue3-sfc-loader'].loadModule('/main.vue', options)
  .then(rootComponent => {

    const app = Vue.createApp(rootComponent)
    app.use(vuetify);
    app.mount(document.body);
  })
  .catch(ex => console.error(ex));

</script>
</body>
</html>
```

**Reply by FranckFreiburger** [reply]
does it meet your needs?
